include/crm_internal.h

```
@@ -198,6 +199,66 @@ crm_set_bit(const char *function, const char *target, long long word, long long
+ * \param _str[inout], the string buffer of size >= \c _max
+ * \param _offset[in], how many bytes from the beginning of \c str to operate
+ * \param _max[in], sentinel to prevent overruns as size <= real \c _str size
+ *
+ * \retval bytesize intended for writing (!= written), 0 on subsequent overrun,
+ *         number of bytes to fill the buffer if \c _str is \c NULL
+ *         or when underlying \c snprintf returned negative value
+ *         (hence when variable behind \c _offset is used as an accumulator
+ *         starting at 0, incremented with values yielded from repeated
+ *         invocations of this macro for a given isolated use case [with
+ *         non-empty desired string production], "no overrun AND \c str OK
+ *         AND no \c snprintf error" is implied by ending up with
+ *         a value >[=] 0 and less than \c max)
+ *
+ * \warning Avoid using parameters with side effects (e.g., <tt>foo++</tt>).
+ *
+ * \parblock
+ * Proven pattern/idiom for using this macro in the context:
+ * \code{.c}
+ * static const int xpath_max = 1024;
+ * char *xpath = calloc(1, xpath_max);
+ * CRM_CHECK(xpath_string != NULL, return -ENOMEM);
+ * offset += crm_snprintf_offset(xpath, offset, xpath_max, "//%s", head);
+ * offset += crm_snprintf_offset(xpath, offset, xpath_max, "/%s", rest);
+ * // change to "offset >= 0" if an empty string production is OK
+ * CRM_LOG_ASSERT(offset > 0 && offset < xpath_max);
+ * \endcode
+ *
+ * Furthermore, there are several handful implications:
+ * - negative value is never returned
+ * - to distinguish \c snprintf error in case <tt>max - \c offset</tt>
+ *   is returned, check \c errno value (provided it was a priori reset)
+ * \endparblock
+ */
+#define crm_snprintf_offset(_str, _offset, _max, ...)                 \
+    (__likely(_max > _offset && _offset >= 0)                         \
+     ? (__likely(_str != NULL)                                        \
```

The ?: operator is bad enough at the best of times, let alone when nested. Please use if-statements. 
Also, convention is to put macros in a 'do {   } while(0)' block, probably best to make this an inline-function though.

You've also swallowed all error handling, why not:
- have the macro/function refuse to accept `offset < 0`
- have the macro/function return the current string length on success
- log an error and return -1 otherwise
- replace 'offset +=' with 'offset =', eg.
    offset = crm_snprintf_offset(data, offset, size, "uname=\"%s\" ", node->uname);